File: crm-ui/src/app/core/services/task.service.ts
import { inject, Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';

import { environment } from '../../../environments/environment';
import { PageResponse } from '../models/page.model';
import { TaskCreateRequest, TaskResponse, TaskUpdateRequest } from '../models/task.model';

const BASE_URL = `${environment.apiUrl}/tasks`;

/** Optional filters/pagination/sorting for {@link TaskService.getTasks}. */
export interface TaskListParams {
  assigneeId?: number;
  completed?: boolean;
  page?: number;
  size?: number;
  /** Spring `sort` syntax, e.g. `'dueDate,asc'`. */
  sort?: string;
}

/**
 * HTTP access for the `/api/tasks` resource (requirements TSK-01 to TSK-06).
 */
@Injectable({ providedIn: 'root' })
export class TaskService {
  private readonly http = inject(HttpClient);

  /**
   * Loads a page of tasks, optionally filtered by assignee or completion
   * status (`GET /api/tasks?assigneeId&completed&page&size`).
   *
   * @param params optional assigneeId/completed/page/size filters
   * @returns the matching page of tasks
   */
  getTasks(params: TaskListParams = {}): Observable<PageResponse<TaskResponse>> {
    let httpParams = new HttpParams();
    if (params.assigneeId !== undefined) {
      httpParams = httpParams.set('assigneeId', params.assigneeId);
    }
    if (params.completed !== undefined) {
      httpParams = httpParams.set('completed', params.completed);
    }
    if (params.page !== undefined) {
      httpParams = httpParams.set('page', params.page);
    }
    if (params.size !== undefined) {
      httpParams = httpParams.set('size', params.size);
    }
    if (params.sort) {
      httpParams = httpParams.set('sort', params.sort);
    }
    return this.http.get<PageResponse<TaskResponse>>(BASE_URL, { params: httpParams });
  }

  /**
   * Creates a new task (`POST /api/tasks`, requirement TSK-01).
   *
   * @param request the new task's field values
   * @returns the created task
   */
  createTask(request: TaskCreateRequest): Observable<TaskResponse> {
    return this.http.post<TaskResponse>(BASE_URL, request);
  }

  /**
   * Replaces a task's editable fields (`PUT /api/tasks/{id}`, requirement
   * TSK-03).
   *
   * @param id the task id to update
   * @param request the new field values
   * @returns the updated task
   */
  updateTask(id: number, request: TaskUpdateRequest): Observable<TaskResponse> {
    return this.http.put<TaskResponse>(`${BASE_URL}/${id}`, request);
  }

  /**
   * Toggles a task's completed state (`PATCH /api/tasks/{id}/toggle`).
   *
   * @param id the task id to toggle
   * @returns the updated task
   */
  toggleTask(id: number): Observable<TaskResponse> {
    return this.http.patch<TaskResponse>(`${BASE_URL}/${id}/toggle`, {});
  }

  /**
   * Deletes a task (`DELETE /api/tasks/{id}`).
   *
   * @param id the task id to delete
   */
  deleteTask(id: number): Observable<void> {
    return this.http.delete<void>(`${BASE_URL}/${id}`);
  }
}

File: crm-service/src/main/java/com/crm/service/TaskService.java
package com.crm.service;

import com.crm.dto.PageResponse;
import com.crm.dto.TaskCreateRequest;
import com.crm.dto.TaskResponse;
import com.crm.dto.TaskUpdateRequest;
import com.crm.entity.Contact;
import com.crm.entity.Deal;
import com.crm.entity.Task;
import com.crm.entity.User;
import com.crm.exception.ResourceNotFoundException;
import com.crm.repository.ContactRepository;
import com.crm.repository.DealRepository;
import com.crm.repository.TaskRepository;
import com.crm.repository.UserRepository;
import java.time.LocalDate;
import java.util.List;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * Business logic for managing {@link Task}s (requirements TSK-01 to TSK-06).
 */
@Service
@RequiredArgsConstructor
@Slf4j
@Transactional(readOnly = true)
public class TaskService {

    private final TaskRepository taskRepository;
    private final ContactRepository contactRepository;
    private final DealRepository dealRepository;
    private final UserRepository userRepository;

    /**
     * Returns a page of tasks, optionally filtered by assignee and/or
     * completion status (requirement TSK-04).
     *
     * @param assigneeId a user id to filter by, or {@code null} for no
     *                   assignee filter
     * @param completed  a completion status to filter by, or {@code null}
     *                   for no completion filter
     * @param pageable   pagination and sort information
     * @return a page of matching tasks, or an empty page if none match
     */
    public PageResponse<TaskResponse> getTasks(Long assigneeId, Boolean completed, Pageable pageable) {
        Specification<Task> spec = Specification.where(null);

        if (assigneeId != null) {
            spec = spec.and((root, query, cb) -> cb.equal(root.get("assignee").get("id"), assigneeId));
        }
        if (completed != null) {
            spec = spec.and((root, query, cb) -> cb.equal(root.get("completed"), completed));
        }

        Page<Task> page = taskRepository.findAll(spec, pageable);
        return PageResponse.from(page.map(TaskResponse::from));
    }

    /**
     * Returns the incomplete tasks assigned to the given user, ordered by
     * due date (earliest first), for the personal dashboard's "My open
     * tasks" widget (requirement TSK-06).
     *
     * @param assigneeId the assignee's user id
     * @return incomplete tasks assigned to this user ordered by due date, or
     *         an empty list if none exist
     */
    public List<TaskResponse> getOpenTasksForAssignee(Long assigneeId) {
        return taskRepository.findByAssigneeIdAndCompletedFalseOrderByDueDateAsc(assigneeId).stream()
                .map(TaskResponse::from)
                .toList();
    }

    /**
     * Creates a new task (requirement TSK-01).
     *
     * @param request the new task's details
     * @return the created task
     * @throws ResourceNotFoundException if {@code assigneeId} does not match
     *         an existing user, or {@code contactId}/{@code dealId} are set
     *         but do not match an existing contact/deal
     */
    @Transactional
    public TaskResponse createTask(TaskCreateRequest request) {
        Task task = new Task();
        applyFields(task, request.getTitle(), request.getDescription(), request.getDueDate(),
                request.getContactId(), request.getDealId(), request.getAssigneeId());

        Task saved = taskRepository.save(task);
        log.info("Created task '{}' (id={})", saved.getTitle(), saved.getId());
        return TaskResponse.from(saved);
    }

    /**
     * Replaces an existing task's editable fields (requirement TSK-03).
     *
     * @param id      the task id
     * @param request the new field values
     * @return the updated task
     * @throws ResourceNotFoundException if no task with this id exists,
     *         {@code assigneeId} does not match an existing user, or
     *         {@code contactId}/{@code dealId} are set but do not match an
     *         existing contact/deal
     */
    @Transactional
    public TaskResponse updateTask(Long id, TaskUpdateRequest request) {
        Task task = taskRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Task " + id + " not found"));

        applyFields(task, request.getTitle(), request.getDescription(), request.getDueDate(),
                request.getContactId(), request.getDealId(), request.getAssigneeId());

        Task saved = taskRepository.save(task);
        log.info("Updated task {}", saved.getId());
        return TaskResponse.from(saved);
    }

    /**
     * Flips a task's completion flag (requirement TSK-02).
     *
     * @param id the task id
     * @return the updated task
     * @throws ResourceNotFoundException if no task with this id exists
     */
    @Transactional
    public TaskResponse toggleCompleted(Long id) {
        Task task = taskRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Task " + id + " not found"));

        task.setCompleted(!task.isCompleted());

        Task saved = taskRepository.save(task);
        log.info("Task {} completed={}", saved.getId(), saved.isCompleted());
        return TaskResponse.from(saved);
    }

    /**
     * Deletes a task.
     *
     * @param id the task id
     * @throws ResourceNotFoundException if no task with this id exists
     */
    @Transactional
    public void deleteTask(Long id) {
        if (!taskRepository.existsById(id)) {
            throw new ResourceNotFoundException("Task " + id + " not found");
        }
        taskRepository.deleteById(id);
        log.info("Deleted task {}", id);
    }

    /**
     * Applies create/update field values to a task entity, resolving its
     * assignee, contact, and deal references.
     *
     * @param task        the task entity to mutate (new or existing)
     * @param title       the task title
     * @param description a free-form description, or {@code null}
     * @param dueDate     the due date, or {@code null}
     * @param contactId   the id of the contact to link, or {@code null} to
     *                    leave the task unlinked from any contact
     * @param dealId      the id of the deal to link, or {@code null} to
     *                    leave the task unlinked from any deal
     * @param assigneeId  the id of the user this task is assigned to
     * @throws ResourceNotFoundException if {@code assigneeId} does not match
     *         an existing user, or {@code contactId}/{@code dealId} are set
     *         but do not match an existing contact/deal
     */
    private void applyFields(Task task, String title, String description, LocalDate dueDate, Long contactId,
            Long dealId, Long assigneeId) {
        task.setTitle(title);
        task.setDescription(description);
        task.setDueDate(dueDate);

        User assignee = userRepository.findById(assigneeId)
                .orElseThrow(() -> new ResourceNotFoundException("User " + assigneeId + " not found"));
        task.setAssignee(assignee);

        if (contactId != null) {
            Contact contact = contactRepository.findById(contactId)
                    .orElseThrow(() -> new ResourceNotFoundException("Contact " + contactId + " not found"));
            task.setContact(contact);
        } else {
            task.setContact(null);
        }

        if (dealId != null) {
            Deal deal = dealRepository.findById(dealId)
                    .orElseThrow(() -> new ResourceNotFoundException("Deal " + dealId + " not found"));
            task.setDeal(deal);
        } else {
            task.setDeal(null);
        }
    }
}
